source/iterator/cdc.go

```diff
@@ -0,0 +1,236 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iterator
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/conduitio-labs/conduit-connector-hubspot/hubspot"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+)
+
+// CDC is an implementation of a CDC iterator for the HubSpot API.
+type CDC struct {
+	hubspotClient *hubspot.Client
+	resource      string
+	bufferSize    int
+	pollingPeriod time.Duration
+	records       chan sdk.Record
+	errC          chan error
+	stopC         chan struct{}
+	position      *Position
+}
+
+// CDCParams is an incoming params for the [NewCDC] function.
+type CDCParams struct {
+	HubSpotClient *hubspot.Client
+	Resource      string
+	BufferSize    int
+	PollingPeriod time.Duration
+	Position      *Position
+}
+
+// NewCDC creates a new instance of the [CDC].
+func NewCDC(ctx context.Context, params CDCParams) (*CDC, error) {
+	cdc := &CDC{
+		hubspotClient: params.HubSpotClient,
+		resource:      params.Resource,
+		bufferSize:    params.BufferSize,
+		pollingPeriod: params.PollingPeriod,
+		records:       make(chan sdk.Record, params.BufferSize),
+		errC:          make(chan error, 1),
+		stopC:         make(chan struct{}, 1),
+		position:      params.Position,
+	}
+
+	if cdc.position == nil || cdc.position.Timestamp == nil {
+		now := time.Now()
+		cdc.position = &Position{
+			Mode:      CDCPositionMode,
+			Timestamp: &now,
+		}
+	}
+
+	if err := cdc.loadRecords(ctx); err != nil {
+		return nil, fmt.Errorf("initial load record: %w", err)
+	}
+
+	go cdc.poll(ctx)
+
+	return cdc, nil
+}
+
+// HasNext returns a bool indicating whether the iterator has the next record to return or not.
+func (c *CDC) HasNext(ctx context.Context) (bool, error) {
+	return len(c.records) > 0, nil
+}
+
+// Next returns the next record.
+func (c *CDC) Next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case <-ctx.Done():
+		return sdk.Record{}, fmt.Errorf("context cancelled: %w", ctx.Err())
+
+	case err := <-c.errC:
+		return sdk.Record{}, fmt.Errorf("async error: %w", err)
+
+	case record := <-c.records:
+		return record, nil
+	}
+}
+
+// poll polls items at the specified time intervals.
+func (c *CDC) poll(ctx context.Context) {
+	ticker := time.NewTicker(c.pollingPeriod)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-c.stopC:
+			return
+
+		case <-ticker.C:
+			if err := c.loadRecords(ctx); err != nil {
+				c.errC <- fmt.Errorf("load records: %w", err)
+			}
+		}
+	}
+}
+
+// loadRecords loads HubSpot items using timestamp-based filtering and search endpoints.
+// The method tries to retrieve items filtering them by updatedAfter or lastmodifieddate.
+func (c *CDC) loadRecords(ctx context.Context) error {
+	// add a millisecond here in order to skip the processed item
```

Comment:
it it possible to have some items with the same timestamp?

Reply:
In case two records are created simultaneously (up to milliseconds).

I'll check what else we can do with this.

Reply:
I mean batch creating, if difference is milliseconds it is ok

Reply:
@maksenius I've tried bulk import supported by HubSpot, and it looks like the createdAt and updatedAt fields are different for each element. So, yeah, even with bulk import, the timestamps are different.

Reply:
cool